File: Generate_Parentheses.cpp
/*22. Generate Parentheses  QuestionEditorial Solution  My Submissions
Total Accepted: 110143 Total Submissions: 273900 Difficulty: Medium
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
Hide Company Tags Google Uber Zenefits
Hide Tags Backtracking String
Hide Similar Problems (M) Letter Combinations of a Phone Number (E) Valid Parentheses
*/



/*use two int to count the remainning left parenthesis(left) and right parenthesis(right)
each step, if left>0, add a left parenthesis
if right >0, add a right parenthesis
append the resulf if left==0 and right==0
*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ret;
        addParenthesis(ret, "", n, 0);
        return ret;
    }
    
    void addParenthesis(vector<string> &ret, string str, int left, int right){
        if(left==0 && right==0){
            ret.push_back(str);
            return;
        }
        if(left>0){
            addParenthesis(ret, str+"(", left-1, right+1);
        }
        if(right > 0){
            addParenthesis(ret, str+")", left, right-1);
        }
    }
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ret = new ArrayList<>();
        addParenthesis(ret, "", n, 0); //just pass string not StringBuilder
        return ret;
    }
    
    public void addParenthesis(List<String> ret, String str, int left, int right){
        if(left==0 && right==0) {
            ret.add(str.toString());
            return;
        }
        if(left>0){
            addParenthesis(ret, str+"(", left-1, right+1);
        }
        if(right>0){
            addParenthesis(ret, str+")", left, right-1);
        }
    }
}

File: Generate_Parentheses.swift
// Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

// For example, given n = 3, a solution set is:

// [
//   "((()))",
//   "(()())",
//   "(())()",
//   "()(())",
//   "()()()"
// ]

// Solution 1: 2 pointer
// 2 pointer for left & right
// add ( when left > 0
// add ) when right > 0
// return until left == right == 0
// 
// Time complexity:Time Complexity : O(\dfrac{4^n}{\sqrt{n}}). Each valid sequence has at most n steps during the backtracking procedure.
// Space Complexity : O(\dfrac{4^n}{\sqrt{n}}), as described above, and using O(n)O(n) space to store the sequence.
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        guard n > 0 else { return [String]() }
        var res = [String]()
        check(&res, "", n, 0)
        return res
    }
    
    private func check(_ res: inout [String], _ str: String, _ left: Int, _ right: Int) {
        if left == 0, right == 0 { 
            res.append(str)
            return 
        }
        
        if left > 0 {
            check(&res, str+"(", left - 1, right + 1)    
        }
        
        if right > 0 {
            check(&res, str+")", left, right - 1)
        }
    }
}